### ISSUE TITLE
pfring_zc_send_pkt error processing

### ISSUE THREAD

**Issue by andyarus** [ntop/PF_RING#227]
I have problem with pfring_zc_send_pkt(queue, pkt_handle, 1);
For some reason (most likely eror in my code) after several days work,
I get error for checking pfring_zc_send_pkt(queue, pkt_handle, 1) < 0
Packet length 1518
But strerror(errsv) equal "Success"
errsv:0
errsv = errno;
It not be such big problem, if it happens only once.
But when it happens for all next packets I get the same problem (pfring_zc_send_pkt(queue, pkt_handle, 1) < 0)
Why this can happen? 1518 (pfring_zc_pkt_buff len) it's ok?
And how I must process this error?
If this error happen what pf ring doing with this packet? Clear buffer?
When after this error I try get new packet pfring_zc_pkt_buff_data(pkt_handle, queue) I really get new packet?
Or I get the same packet?
I don't understand why:
- packet length 1518 happen
- after first error, other packets don't sending too.

**Comment by cardigliano** [comment]
Please print the pfring_zc_send_pkt() return value and errno and provide them.
What is the buffer_len provided to pfring_zc_create_cluster()?

**Comment by andyarus** [comment]
pfring_zc_send_pkt() return value: -1
errno: 0
strerror:Success
buffer_len:1600 (first I try 1532, then increased to 1600)

Now I get the same problem after hour work.
pkt->len:637 errsv:0 strerror:Success
I add additional log info and see that it's problem only in one thread. Only for one RSS queue for one direction. Other threads works for all other RSS queues in both directions...
Looks like I have problem somewhere in my code. But how detect? Can I get more informative error message from pf_ring?

**Comment by cardigliano** [comment]
Please provide the interface model and the driver you are using.
Is the link up when you get the error? Please check ethtool <interface>

**Comment by andyarus** [comment]
Quad Port Fiber 10 Gigabit Ethernet Bypass Server Adapter **Intel® 82599ES** Based
ixgbe-4.1.5-zc/src
Link is up because other RSS queues work in the same time

**Comment by cardigliano** [comment]
Any message in dmesg?

**Comment by andyarus** [comment]
Nothing unusual

**Comment by andyarus** [comment]
I think it's problem in my code. I will look for.
I have another question about jumbo frame (raised the issue buffer len)
If I need processing jumbo frame and my card support jumbo frame up to 16KB
I must provide buffer_len for pfring_zc_create_cluster equal 16384? But it consume much memory for hugepages...
What's PF RING way processing jumbo frames?

**Comment by cardigliano** [comment]
@andyarus if you use jumbo frames, ZC needs to allocate buffers large enough to contains jumbo frames, consuming much more memory. That's expected.

**Comment by andyarus** [comment]
What's happen if I set buffer_len for pfring_zc_create_cluster equal 1600 and on NIC come jumbo frame? He will be dropped?

I looked more logs and saw
rc:-1
errsv:107 strerror:Transport endpoint is not connected

**Comment by cardigliano** [comment]
pfring_zc_open_device() should fail if you try to open a ZC interface with an MTU smaller than the cluster buffer len. Are you using a ZC interface (zc:ethX) or a standard one?

**Comment by andyarus** [comment]
ZC
ifconfig say interface MTU:1500
pfring_zc_create_cluster buffer_len:1600
I open 16 RSS queues per interface without errors. All working good for time but than begins problems with sending packets to specific RSS queue, other RSS queues working in the same time. I think I have problem in my code with pf ring packet buffer memory corruption or something like this

**Comment by andyarus** [comment]
Again get the same error after several hours work.
But without "Transport endpoint is not connected" error.
rc:-1 errsv:0 strerror:Success
Problem starting from packet length 1518
But in other thread (direction the same). Accordingly in other RSS queue.

Before I use pfring_zc_run_balancer for distribution packets between processor cores.
But I see that such software distribution slow down packet processing. About 200 microseconds higher ping for ICMP packets.
Now I use 16 RSS queues per interface for both directions (RX/TX).
pfring_zc_open_device for every queue
I simultaneously read and write from/to this queues in different threads.
Can I have problem because of this?
It's ok read and write to the same RSS queue without blocking (mutex)?
What's difference between RSS queue opening for read (rx_only) and RSS queue opening for write (tx_only)?

**Comment by cardigliano** [comment]
@andyarus please note both queues and devices are single-producer/single-consumer, in essence you cannot use the same queue from multiple threads.
As of rx_only vs tx_only, you should open a queue in rx_only mode for capturing traffic, and tx_only for transmitting traffic. You can open the same queue in rx_only mode using one thread, and tx_only mode using another thread.

**Comment by andyarus** [comment]
Thanks!
I capture and transmit traffic from different threads. For every thread own queue.
But I open all queues from main thread and then pass the pointer in working thread who binding to other core. For capture only.
For transmit I use global array with pointers to open queues. Every thread have own index.
It's wrong? I must open queue from working thread after binding to specific core in order to memory allocated from the same processor?

```
typedef struct my_pkt_struct {
    pfring_zc_pkt_buff *buffer;
    u_char *packet;

    u_int32_t len;
    some_struct_t *struct_p;
    u_int8_t var2;
    u_int32_t var3;
    u_int16_t var4;
} my_pkt_struct_t;

typedef struct some_struct_td {
    pfring_zc_pkt_buff *buffer;
    pfring_zc_queue *queue;
    ...
} some_struct_td_t;

void *working_thread(void *_td)
{
      some_struct_td_t *td = _td;
      bind2core(_td->core); // the same core for RX thread with RX queue and TX thread for TX queue

      while (run) {
          if (pfring_zc_recv_pkt(td->queue, &td->buffer, 1) > 0) {
              my_pkt_struct_t *pkt = (my_pkt_struct_t *) malloc(sizeof(my_pkt_struct_t));
              pkt->packet = pfring_zc_pkt_buff_data(td->buffer, td->queue);
              pkt->len = td->buffer->len;
              pkt->struct_p = NULL;
              pkt->var2 = 0;
              pkt->var3 = 0;
              pkt->buffer = td->buffer;
              pkt->var4 = 0;

              // some processing
              nDPI protocol detection on pkt->packet
              // some processing

              free(pkt);
          }
      }
}
```

Now I did memory allocation for every packet (my_pkt_struct_t) and while all working good (day)

When I try allocate memory once and use it again, I get described above problem

```
void *working_thread(void *_td)
{
      some_struct_td_t *td = _td;
      bind2core(_td->core); // the same core for RX thread with RX queue and TX thread for TX queue

      my_pkt_struct_t pkt;

      while (run) {
          if (pfring_zc_recv_pkt(td->queue, &td->buffer, 1) > 0) {
              pkt.packet = pfring_zc_pkt_buff_data(td->buffer, td->queue);
              pkt.len = td->buffer->len;
              pkt.struct_p = NULL;
              pkt.var2 = 0;
              pkt.var3 = 0;
              pkt.buffer = td->buffer;
              pkt.var4 = 0;

              // some processing
              nDPI protocol detection on pkt.packet
              // some processing
          }
      }
}
```

It's seems better allocate memory once and use again rather than allocate and than free memory for every packet.
But I get problem with pfring_zc_send_pkt(tx_queues_some_direction[td->thread_id], &pkt->buffer, 1)

But question what's happen and where?
It's possible memory corruption with pkt.packet when nDPI protocol detect?

I don't do copy like this

```
u_char *packet_check = (u_char *) malloc(pkt->len);
memcpy(packet_checked, pkt->packet, pkt->len);
```

I try detect protocol on pkt->packet right away.
I think nDPI reliable and I don't need this unnecessary copy.

Or somewhere else in my code I corrupt pf ring packet buffer data.
But what I can do so wrong that pfring_zc_send_pkt starting return error for some one thread after several days (or hours) of successful work?

**Comment by andyarus** [comment]
I don't do error processing for pfring_zc_pkt_buff_data...
What's happen if it return NULL: crash or my problem with sending

**Comment by cardigliano** [comment]
@andyarus in essence if you memcpy the packet after receiving it, everything works. What are you doing with the packet besides ndpi? I do not see where you are sending it for instance.

**Comment by andyarus** [comment]
Now app still working starting from 20.06.2017.
Nope. I don't memcpy packet data.
I allocate memory for my structure. In structure I have pointer
`u_char *packet;`
And if for every processing packet I allocate memory for structure that all working.
If I allocate structure once per thread and use this memory for every packet, I get problem with sending. But why...
In both cases I get pointer from pf ring
`pkt->packet = pfring_zc_pkt_buff_data(td->buffer, td->queue);`
I do many bad things with packet :-) So bad that I can not talk about it :-)
In essence just packet processing, DPI, adding to queue (or immediately sending), modification (rarely) and in the end sending to RSS queue
```
if ((rc = pfring_zc_send_pkt(global_array_with_pointers_to_open_RSS_TX_queues[td->thread_id], &pkt->buffer, 1)) < 0) {
}
```

**Comment by andyarus** [comment]
What if I need drop packet?
I get packet from queue
`if (pfring_zc_recv_pkt(td->queue, &td->buffer, 1) > 0) {`
If then I send this packet to TX queue, pf ring free this memory? Or just will know that this memory can be reused.
Buf if I get from queue and do nothing with this packet.
What happens? This memory will be reused later?

**Comment by andyarus** [comment]
Again get the same problem with sending

**Comment by andyarus** [comment]
rc:-1 pkt->len:64 errsv:0 strerror:Success
rc:-1 pkt->len:1263 errsv:0 strerror:Success
rc:-1 pkt->len:70 errsv:0 strerror:Success
rc:-1 pkt->len:1258 errsv:0 strerror:Success
rc:-1 pkt->len:1258 errsv:0 strerror:Success
rc:-1 pkt->len:1518 errsv:0 strerror:Success
rc:-1 pkt->len:1518 errsv:0 strerror:Success
rc:-1 pkt->len:70 errsv:0 strerror:Success
rc:-1 pkt->len:64 errsv:0 strerror:Success
...
1000 this errors counting and I exit app

Seems it's happen with packet, that I add to queue before.
It's possible that while packet was in queue pf ring rewrite buffer for this packet and when I try sending there already were other data?

```
# Max number of RX slots
ethtool -G $IF rx 65535

# Max number of TX slots
ethtool -G $IF tx 65535
```

What's value I need to set there?

```
pfring_zc_create_cluster
tot_num_buffers	The total number of buffers to reserve for queues/devices/extra allocations.

#define QUEUE_LEN                  32768
#define MAX_CARD_SLOTS             65536

int tot_num_buffers = 
    (
        (
            (
            one_direction_interfaces_count + other_direction_interfaces_count
            ) * 2
        ) * MAX_CARD_SLOTS
        +
        (
            (
            one_direction_interfaces_count + other_direction_interfaces_count
            ) * 2 // IN OUT per interface
        ) * threads_count * QUEUE_LEN
    );

```
I allocate enough buffers?
How to correctly calculate this value?
What's value I must set for **Max number of RX/TX slots**?

**Comment by andyarus** [comment]
```
sudo ethtool -g ethX

Pre-set maximums:
RX:		32768
RX Mini:	0
RX Jumbo:	0
TX:		32768
Current hardware settings:
RX:		32768
RX Mini:	0
RX Jumbo:	0
TX:		32768
```

Maximum possible value 32768?
So it makes no sense set...
ethtool -G $IF rx 65535
ethtool -G $IF tx 65535

**Comment by cardigliano** [comment]
> pkt->packet = pfring_zc_pkt_buff_data(td->buffer, td->queue);
> In essence just packet processing, DPI, adding to queue (or immediately sending), modification (rarely) and in the end sending to RSS queue

When you get a packet pointer using pfring_zc_pkt_buff_data(), that pointer is tied to td->buffer, thus you cannot enqueue or keep aside the packet pointer for later processing after reusing td->buffer.


> I get packet from queue
> if (pfring_zc_recv_pkt(td->queue, &td->buffer, 1) > 0) {
> If then I send this packet to TX queue, pf ring free this memory? Or just will know that this memory can be reused.
> Buf if I get from queue and do nothing with this packet.
> What happens? This memory will be reused later?

When you send a packet td->buffer, after calling pfring_zc_send_pkt() td->buffer points to a new buffer returned by pfring_zc_send_pkt() and can be used to call pfring_zc_recv_pkt() again. Instead, if you are not sending td->buffer, it can be reused to call pfring_zc_recv_pkt() again. In essence in any case you can reuse td->buffer.

**Comment by andyarus** [comment]
```
if (pfring_zc_recv_pkt(td->queue, &td->buffer, 1) > 0) {
    ...
    pkt->packet = pfring_zc_pkt_buff_data(td->buffer, td->queue);
    pkt->buffer = td->buffer;
    ...
}
```

I enqueue pkt to queue for sending later.
`pkt->buffer = td->buffer;`
It's possible that
`pfring_zc_recv_pkt(td->queue, &td->buffer, 1)`
return pointer to buffer equal pkt->buffer from my queue?
And when I try send pkt->buffer from my queue, there will already be another packet?
Because of this I get an error when sending?!

**Comment by andyarus** [comment]
What's pf ring way QoS(quality of service)?
I need enqueue packet for sending later.
How do I do it correctly?!

**Comment by cardigliano** [comment]
> I enqueue pkt to queue for sending later.
> pkt->buffer = td->buffer;
> It's possible that
> pfring_zc_recv_pkt(td->queue, &td->buffer, 1)
> return pointer to buffer equal pkt->buffer from my queue?
> And when I try send pkt->buffer from my queue, there will already be another packet?
> Because of this I get an error when sending?!

After calling pfring_zc_recv_pkt(td->queue, &td->buffer, 1) td->buffer points to the received buffer, and the old td->buffer is no longer available as it can be reused by the queue (if you previously assign it to pkt->buffer, and later you access pkt->buffer, you are doing a wrong operation).

**Comment by cardigliano** [comment]
> What's pf ring way QoS(quality of service)?
> I need enqueue packet for sending later.
> How do I do it correctly?!

If you want to keep buffers aside, you need to allocate more buffers to be provided to pfring_zc_recv_pkt(), you can do that with pfring_zc_get_packet_handle() (from a global resource) or with pfring_zc_get_packet_handle_from_pool() (from a pool resource, to be used for runtime allocations from thread, with one pool per thread in order to avoid mutexes) 
You can create a pfring_zc_buffer_pool calling pfring_zc_create_buffer_pool(<number of buffers in the pool>)

**Comment by andyarus** [comment]
Thanks! Now I understand more clearly, but not all.
Can you please give a example of code for this situation?

```
typedef struct my_pkt_struct {
    pfring_zc_pkt_buff **buffer;
    ...
} my_pkt_struct_t;

void *working_thread(void *_td)
{
      some_struct_td_t *td = _td;
      bind2core(_td->core); // the same core for RX thread with RX queue and TX thread for TX queue

      u_int32_t pool_len = 100500;
      pfring_zc_buffer_pool *buf_pool = pfring_zc_create_buffer_pool(td->zc, pool_len); // td->zc == cluster

      pfring_zc_pkt_buff *buffer = pfring_zc_get_packet_handle_from_pool(buf_pool);

      my_pkt_struct_t pkt;

      while (run) {
          if (pfring_zc_recv_pkt(td->queue, &buffer, 1) > 0) {
              
              pkt.packet = pfring_zc_pkt_buff_data(buffer, td->queue);
              pkt.len = buffer->len;
              pkt.struct_p = NULL;
              pkt.var2 = 0;
              pkt.var3 = 0;
              pkt.buffer = &buffer;
              pkt.var4 = 0;

              // some processing
              nDPI protocol detection on pkt.packet
              // some processing              

              if (...) {
                  // dequeue pkt from my queue
                  my_queue_t *item = my_queue;
                  pfring_zc_release_packet_handle_to_pool(buf_pool, *item->pkt->buffer);
                  // send pkt
                  if ((rc = pfring_zc_send_pkt(global_array_with_pointers_to_open_RSS_TX_queues[td->thread_id], pkt->buffer, 1)) < 0) {
                  }
              } else {
                  // enqueue pkt to my queue
                  buffer = pfring_zc_get_packet_handle_from_pool(buf_pool);
              }
          }
      }
}

```
It's wrong?
How to correctly do what you described.

First I create pool and get buffer from this pool
`pfring_zc_pkt_buff *buffer = pfring_zc_get_packet_handle_from_pool(buf_pool);`

Then I use this buffer for packet receiving
`if (pfring_zc_recv_pkt(td->queue, &buffer, 1) > 0) {`

Now I need enqueue packet for sending later. What's I need to do?
Like this?
`buffer = pfring_zc_get_packet_handle_from_pool(buf_pool);`

Now I have new buffer for packet receiving? Or not?
When I should release pkt.buffer? When dequeue packet from my queue and (important) before sending?
I can't understand how to do it right.

**Comment by andyarus** [comment]
`if (pfring_zc_recv_pkt(td->queue, &buffer, 1) > 0) {`

pfring_zc_recv_pkt make swap? Right?
It append to device queue free buffer and get from this queue buffer with data?

```
if ((rc = pfring_zc_send_pkt(global_array_with_pointers_to_open_RSS_TX_queues[td->thread_id], pkt->buffer, 1)) < 0) {
}
```

pfring_zc_send_pkt if sent was success get free buffer (from where? from some global queue with free buffers? Or from some pool buffers?)
If sent was fail, then what?
Whatever pfring_zc_send_pkt return new free buffer? (which would point pkt->buffer)

If I need send 10 packets at once

Sent the first packet
pkt->buffer point to the new buffer A
Sent the second packet
pkt->buffer point to the new buffer B
Sent the third packet
pkt->buffer point to the new buffer С

But what happens with buffer A and B? They will be lost?

I need manually return this buffer back to pool in this situation like this?
```
pfring_zc_release_packet_handle_to_pool(buf_pool, A);
pfring_zc_release_packet_handle_to_pool(buf_pool, B);
```

and so on except last returned buffer that I will use in the next time when I will recieve new packet from queue

Sent the last packet from my queue
pkt->buffer point to the new buffer G

buffer = G

`if (pfring_zc_recv_pkt(td->queue, &buffer, 1) > 0) {`

It's make swap: append G to end device queue and return me buffer with data from this queue?

It's correct?

**Comment by cardigliano** [comment]
> if (pfring_zc_recv_pkt(td->queue, &buffer, 1) > 0) {
> pfring_zc_recv_pkt make swap? Right?
> It append to device queue free buffer and get from this queue buffer with data?

Correct

> if ((rc = pfring_zc_send_pkt(global_array_with_pointers_to_open_RSS_TX_queues[td->thread_id], pkt->buffer, 1)) < 0) {
> }
> pfring_zc_send_pkt if sent was success get free buffer (from where? from some global queue with free buffers? Or from some pool buffers?)

It gives the buffer to send to the queue, and gets a fresh buffer from the queue (which is initialised with buffers when you create it)

> If sent was fail, then what?

If it fails, the swap doesn't happen: you still have your buffer

> Buf what happens with buffer A and B? They will be lost?

You should provide them to the next pfring_zc_recv_pkt(), or return them to a pool (pfring_zc_release_packet_handle_to_pool) in case you do not need to call pfring_zc_recv_pkt()

> buffer = G
> if (pfring_zc_recv_pkt(td->queue, &buffer, 1) > 0) {
> It's make swap: append G to end device queue and return me buffer with data from from this queue?

Correct, it provides G to the queue to be used for future packets, and get a buffer with data.

**Comment by andyarus** [comment]
```
u_int32_t pool_len = 1000;
pfring_zc_buffer_pool *buf_pool = pfring_zc_create_buffer_pool(td->zc, pool_len);

pfring_zc_pkt_buff *buffer = pfring_zc_get_packet_handle_from_pool(buf_pool);

pfring_zc_pkt_buff *buffer_previous = buffer;

// td->queue RX queue
if (pfring_zc_recv_pkt(td->queue, &buffer, 1) > 0) {
```

Possible options:
1. buffer_previous != buffer
It's normal? Free buffer (buffer_previous) was appended to the td->queue (front or end?)
and from queue a buffer (buffer) with packet data was obtained
Or
```
pfring_zc_release_packet_handle(cluster, buffer_previous); // Release a buffer to global resources
*buffer = get packet from RX queue
```
How right?

2. buffer_previous == buffer
It's not normal?
Why it's happen? What's it's mean?! And what to do?





--------------------------------------------------------------------------------------------------------

```
typedef struct my_pkt_struct {
    pfring_zc_pkt_buff **buffer;
    ...
} my_pkt_struct_t;

pfring_zc_pkt_buff *buffer_previous = *pkt->buffer;
if ((rc = pfring_zc_send_pkt(global_array_with_pointers_to_open_RSS_TX_queues[td->thread_id], pkt->buffer, 1)) < 0) {
}
```

Possible options:
1. buffer_previous != *pkt->buffer
It's ok
Buffer with data was appended to the TX device queue for sending
pfring_zc_send_pkt assign *pkt->buffer to the new buffer **obtained from where**?
From where pfring_zc_send_pkt get new free buffer?!
Allocate a buffer from global resources? 
`*pkt->buffer = pfring_zc_get_packet_handle(cluster);`

2. buffer_previous == *pkt->buffer
It's not normal?
Why it's happen? What's it's mean?! And what to do?

**Comment by andyarus** [comment]
Got it!
I was doing wrong
In func
`queue_packets_processing(pfring_zc_pkt_buff *buffer, ...) {`
after enqueue packet to my queue to send later
I did
```
pfring_zc_pkt_buff* tmp_buf = pfring_zc_get_packet_handle_from_pool(buf_pool);
if (tmp_buf == NULL) {
... 
} else {
    buffer = tmp_buf;
}
```
And after return from function buffer not changed!
As a result, I used the same buffer that I added to my queue to send later.
Therefore I get problem:
buffer_previous == buffer

Remade
```
queue_packets_processing(pfring_zc_pkt_buff **buffer, ...) {
*buffer = tmp_buf;
```
It seems now everything works well. New buffers not equal previous.
